addons/loyalty/models/loyalty.py

```diff
@@ -0,0 +1,70 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import fields, models, api, _
+from odoo.exceptions import ValidationError
+
+
+class LoyaltyProgram(models.Model):
+    _name = 'loyalty.program'
+    _description = 'Loyalty Program'
+
+    name = fields.Char(string='Loyalty Program Name', index=True, required=True, translate=True, help="An internal identification for the loyalty program configuration")
+    points = fields.Float(string='Point per $ spent', help="How many loyalty points are given to the customer by sold currency")
+    rule_ids = fields.One2many('loyalty.rule', 'loyalty_program_id', string='Rules')
+    reward_ids = fields.One2many('loyalty.reward', 'loyalty_program_id', string='Rewards')
+    active = fields.Boolean(default=True)
+
+
+class LoyaltyRule(models.Model):
+    _name = 'loyalty.rule'
+    _description = 'Loyalty Rule'
+
+    name = fields.Char(index=True, required=True, help="An internal identification for this loyalty program rule")
+    loyalty_program_id = fields.Many2one('loyalty.program', string='Loyalty Program', help='The Loyalty Program this exception belongs to')
+    points_quantity = fields.Float(string="Points per Unit")
+    points_currency = fields.Float(string="Points per $ spent")
+    rule_domain = fields.Char()
+
+
+class LoyaltyReward(models.Model):
+    _name = 'loyalty.reward'
+    _description = 'Loyalty Reward'
+
+    name = fields.Char(index=True, required=True, help='An internal identification for this loyalty reward')
+    loyalty_program_id = fields.Many2one('loyalty.program', string='Loyalty Program', help='The Loyalty Program this reward belongs to')
+    minimum_points = fields.Float(help='The minimum amount of points the customer must have to qualify for this reward')
+    reward_type = fields.Selection([('gift', 'Free Product'), ('discount', 'Discount')], required=True, help='The type of the reward', default="gift")
+    gift_product_id = fields.Many2one('product.product', string='Gift Product', help='The product given as a reward')
+    point_cost = fields.Float(string='Reward Cost', help="If the reward is a gift, that's the cost of the gift in points. If the reward type is a discount that's the cost in point per currency (e.g. 1 point per $)")
+    discount_product_id = fields.Many2one('product.product', string='Discount Product', help='The product used to apply discounts')
```

Comment:
Something strange with this field when trying to add a reward to the website loyalty, see 
![image](https://user-images.githubusercontent.com/30048408/117440910-fb0c9c80-af34-11eb-8eb6-30920fcbe897.png)
Why do I have to specify a `Discount Product`? Shouldn't the system create it silently for me as in coupon?